include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,802 @@
+
+                for (const sidb_cluster_state_composition& composition : m.compositions)
+                {
+                    for (const sidb_cluster_state& cst : composition)
+                    {
+                        if (cst.internal_pot_bounds.count(sidb_ix) != 0)
+                        {
+                            take_meet_of_potential_bounds<bound_direction::LOWER>(
+                                lb_meet, cst.get_pot_bound<bound_direction::LOWER>(sidb_ix));
+                            take_meet_of_potential_bounds<bound_direction::UPPER>(
+                                ub_meet, cst.get_pot_bound<bound_direction::UPPER>(sidb_ix));
+                        }
+                    }
+                }
+
+                add_pot_proj(parent, sidb_ix, potential_projection{lb_meet, static_cast<uint64_t>(m)});
+                add_pot_proj(parent, sidb_ix, potential_projection{ub_meet, static_cast<uint64_t>(m)});
+            }
+        }
+    }
+
+    void move_up_hierarchy() noexcept
+    {
+        if (clustering.size() == 1)
+        {
+            terminate = true;
+            return;
+        }
+
+        // find the parent with the minimum cluster size
+        const sidb_cluster_ptr& min_parent =
+            (*std::min_element(clustering.cbegin(), clustering.cend(),
+                               [](const sidb_cluster_ptr& c1, const sidb_cluster_ptr& c2)
+                               { return c1->get_parent()->size() < c2->get_parent()->size(); }))
+                ->get_parent();
+
+        for (const sidb_cluster_ptr& c : min_parent->children)
+        {
+            clustering.erase(c);
+        }
+
+        derive_children_recv_bounds_without_siblings(min_parent);
+
+        construct_merged_charge_state_space(min_parent);
+
+        if (min_parent->charge_space.empty())
+        {
+            terminate = true;
+            return;
+        }
+
+        construct_merged_potential_projections(min_parent);
+
+        compute_meets_for_internal_pot_bounds(min_parent);
+
+        clustering.emplace(min_parent);
+
+        update_charge_spaces(min_parent->uid);
+    }
+
+    constexpr inline uint64_t maximum_top_level_multisets(const uint64_t number_of_sidbs) const noexcept
+    {
+        // computes nCr(N + 2, 2)                             // computes nCr(N + 1, 1)
+        return base == 3 ? ((number_of_sidbs + 1) * (number_of_sidbs + 2)) / 2 : number_of_sidbs + 1;
+    }
+
+    const uint8_t base;
+
+    const sidb_cluster_ptr top_cluster;
+
+    sidb_clustering clustering{};
+
+    bool terminate = false;
+
+    const uint64_t witness_partitioning_max_cluster_size;
+
+    const std::array<double, 4> mu_bounds_with_error;
+};
+
+}  // namespace detail
+
+/**
+ * The purely constructive *Ground State Space* algorithm is the key ingredient of the *ClusterComplete* exact SiDB
+ * simulator that revolutionizes the domain of exact SiDB simulation. It uses iterative "loop until fixpoint" concepts
```

Comment:
In fact, it revolutionizes the domain 😁, but try to write the docstring more "neutral".

Reply:
Thanks for the feedback! I'm learning there's a difference between paper-writing and docstring-writing